### PR TITLE
Support alternative deployment method split across infra and tenant

### PR DESCRIPTION
**Pull request by davidvossel**
This adds support for deploying the csi driver in a way that does not give the tenant cluster any API access to the infra cluster. Only the node daemonset is deployed into the tenant cluster which does not need to talk to the infra cluster api at all. The csi driver's controller deployment is placed in the infra cluster with a kubeconfig that allows accessing the tenant cluster.

```release-note
Support deployment method that splits controller to infra cluster and node ds to tenant
```

### CODE REVIEW

File: deploy/split-infra-tenant/040-controller.yaml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  replicas: 1
+  selector:
+    matchLabels:
+      app: kubevirt-csi-driver
+  template:
+    metadata:
+      labels:
+        app: kubevirt-csi-driver
+    spec:
+      hostNetwork: true
```
Comment (awels) [inline deploy/split-infra-tenant/040-controller.yaml:15]: do we really need `hostNetwork: true`?

File: deploy/split-infra-tenant/030-node.yaml [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    matchLabels:
+      app: kubevirt-csi-driver
+  updateStrategy:
+    type: RollingUpdate
+  template:
+    metadata:
+      labels:
+        app: kubevirt-csi-driver
+    spec:
+      hostNetwork: true
```
Comment (awels) [inline deploy/split-infra-tenant/030-node.yaml:18]: do we really need `hostNetwork: true`?

File: deploy/split-infra-tenant/020-autorization.yaml [hunk 3]
```diff
@@ -0,0 +1,134 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: kubevirt-csi-node-sa
+  namespace: kubevirt-csi-driver
+---
+apiVersion: v1
```
Comment (awels) [inline deploy/split-infra-tenant/020-autorization.yaml:7]: Maybe split these into two files? one for the infra and one for the tenant. The infra doesn't need the node rbac and the tenant doesn't need the controller rbac.

File: deploy/split-infra-tenant/000-namespace.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: kubevirt-csi-driver
+  annotations:
```
Comment (awels) [inline deploy/split-infra-tenant/000-namespace.yaml:5]: probably don't need this annotation.

File: deploy/infra-cluster-service-account.yaml [hunk 5]
```diff
@@ -29,4 +29,4 @@ roleRef:
 subjects:
 - kind: ServiceAccount
   name: kubevirt-csi
----
\ No newline at end of file
+---
```
Comment (awels) [inline deploy/infra-cluster-service-account.yaml:32]: I think you can remove this file, you split the rbac into two below.

**Review by awels — COMMENTED** [review]
I know this is just a first step, but we should really split the cmd/kubevirt-csi-driver.go into two pieces, the controller and the node. For one with the node, you can pass the node id into the node server from the pod, instead of having to look it up with a kube client.

File: deploy/split-infra-tenant/040-controller.yaml (hunk 1 above)
Comment (davidvossel) [reply deploy/split-infra-tenant/040-controller.yaml:15]: Ha, i didn't catch that. Great question. I copied these over from the other deploy scripts and just made a few modifications here and there. I'll try taking this hostNetwork part out and see what happens.

File: deploy/split-infra-tenant/020-autorization.yaml (hunk 3 above)
Comment (davidvossel) [reply deploy/split-infra-tenant/020-autorization.yaml:7]: yep, makes sense

File: deploy/infra-cluster-service-account.yaml (hunk 5 above)
Comment (davidvossel) [reply deploy/infra-cluster-service-account.yaml:32]: this is still needed for the use case where both controller and node are deployed in tenant cluster.

File: deploy/split-infra-tenant/040-controller.yaml (hunk 1 above)
Comment (davidvossel) [reply deploy/split-infra-tenant/040-controller.yaml]: fixed, we don't need hostNetwork. I think this was left over from copying manifests from the aws csi driver's yaml.

File: deploy/split-infra-tenant/020-autorization.yaml (hunk 3 above)
Comment (davidvossel) [reply deploy/split-infra-tenant/020-autorization.yaml]: fixed

**Comment by davidvossel** [comment]
updated

> I know this is just a first step, but we should really split the cmd/kubevirt-csi-driver.go into two pieces, the controller and the node. For one with the node, you can pass the node id into the node server from the pod, instead of having to look it up with a kube client.

yep, there's some refactoring to be done here to better separate the two code paths. For delivery, we need to keep this down to one binary and one container image though.

**Comment by awels** [comment]
I can see the single container (one less build), don't quite get the single binary though, nothing stopping you from putting 2 binaries in one container. The arguments and code paths for the controller and the node ds are quite different, so it makes little sense to me to try and put both in a single binary.

**Comment by awels** [comment]
Note none of that for this PR, just wanted to get the discussion on it started since I think it will be important in the long run.
/lgtm

**Comment by davidvossel** [comment]
> I can see the single container (one less build), don't quite get the single binary though, nothing stopping you from putting 2 binaries in one container. The arguments and code paths for the controller and the node ds are quite different, so it makes little sense to me to try and put both in a single binary.

I see the advantages as a smaller image and a single entrypoint. but yeah we can discuss this more later.

**Comment by brybacki** [comment]
@davidvossel @awels do we want to have a test lane for the split deployment?

**Comment by awels** [comment]
Yeah we probably want to run the tests in both modes.

**Pull request merged**